### DISCUSSION TITLE
Electron

### DISCUSSION THREAD

**Discussion by tripower** [MarkusJx/node-java-bridge#32] [Q&A] [answered]
Hello

will this bridge also work with Electron?
(without Typescript)
and which versions are supported?
of Electron/ NodeJS

greetings
Klaus

**Comment by MarkusJx** [comment, marked as answer]
Hi

Electron _should_ be supported. At least v1 did. This was actually kind of the point for making this in the first place. But I've never tried v2 of this addon against electron. According to the electron docs, everything from v15 and up should be supported as these versions run on node 16, which is the minimum requirement for this package.

But I can check if electron works later today, I'll keep you posted.

cheers

**Reply by tripower** [reply]
sounds good
it should work with Electron v21 and so NideJS v16.xxx
would be fine if you could test it.
thanks

if you need example code for integration
I found this which works also with E. v21
https://github.com/serialport/electron-serialport

greetings
Klaus

**Reply by MarkusJx** [reply]
Seems to be working with electron 21 and node 16

**Reply by tripower** [reply]
thats cool
thanks for the test
would be fine if you  could adapt your doc with the info and also some version infos of the tools are needed for build
thx

greetings
Klaus

**Reply by tripower** [reply]
@MarkusJx do I need all the tools and a build 
or is on Windows the install enough0ü?

**Reply by MarkusJx** [reply]
No, you don't need to install anything, maybe some C++ Redistributables (maybe VS13, VS15 and 17 should be fine) but all of those should be installed on a _normal_ system

**Reply by tripower** [reply]
ok. sounds good. I will try it
thanks for the quick response

**Reply by tripower** [reply]
@MarkusJx any ideas?
because our the import brings an error 
**const { importClass }= require('java-bridge')**

import cannot be used because we do not user Typescript

Can you checkin your running example to github or dl in googledrive?
or does your module not work with "normal" NodeJS and needs Typescript??

thanks in advance
cheers Klaus

**Reply by MarkusJx** [reply]
Using the [electron-quick-start](https://github.com/electron/electron-quick-start) example, I've enabled node integration in `main.js` and added some test code to `preload.js`.

<details>
<summary>main.js</summary>

```js
// Modules to control application life and create native browser window
const {app, BrowserWindow} = require('electron')
const path = require('path')
const {importClass} = require('java-bridge')

function createWindow () {
  // Create the browser window.
  const mainWindow = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      preload: path.join(__dirname, 'preload.js'),
      nodeIntegration: true,
    }
  })

  // and load the index.html of the app.
  mainWindow.loadFile('index.html')

  // Open the DevTools.
  // mainWindow.webContents.openDevTools()
  const System = importClass('java.lang.System');
  System.out.printlnSync('Hello from main through java');
}

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.whenReady().then(() => {
  createWindow()

  app.on('activate', function () {
    // On macOS it's common to re-create a window in the app when the
    // dock icon is clicked and there are no other windows open.
    if (BrowserWindow.getAllWindows().length === 0) createWindow()
  })
})

// Quit when all windows are closed, except on macOS. There, it's common
// for applications and their menu bar to stay active until the user quits
// explicitly with Cmd + Q.
app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') app.quit()
})

// In this file you can include the rest of your app's specific main process
// code. You can also put them in separate files and require them here.
```

</details>

<details>
<summary>preload.js</summary>

```js
/**
 * The preload script runs before. It has access to web APIs
 * as well as Electron's renderer process modules and some
 * polyfilled Node.js functions.
 * 
 * https://www.electronjs.org/docs/latest/tutorial/sandbox
 */
window.addEventListener('DOMContentLoaded', () => {
  const replaceText = (selector, text) => {
    const element = document.getElementById(selector)
    if (element) element.innerText = text
  }

  for (const type of ['chrome', 'node', 'electron']) {
    replaceText(`${type}-version`, process.versions[type])
  }
});

// This is new
const {importClass} = require('java-bridge');

const System = importClass('java.lang.System');

// Prints out 'Hello world from Java!' in the cmd
System.out.printlnSync('Hello world from Java!');
```

</details>

Starting the program using `npm start` prints out 'Hello from main through java' and 'Hello world from Java!' in the cmd just as expected.

**Reply by tripower** [reply]
thats perfect🤗
thank you very much!
so hopefully here also somebody says "Hello" 😂👍

greetings
Klaus

**Reply by tripower** [reply]
@MarkusJx 
Status

at the moment the app crashes here
const System = importClass('java.lang.System'); 

npm ERR! code ELIFECYCLE 
npm ERR! errno 3221225477 
....

java should be available
PS D:\workspace\electron_example\electron-quick-start> java -version openjdk version "11.0.17" 2022-10-18 OpenJDK Runtime Environment Temurin-11.0.17+8 ...

some additional version infos
Oracle Java 32bit
NodeJS 16.15.0  32bit
Electron 21.2.3
Windows 10

any ideas?

greetings Klaus

**Reply by MarkusJx** [reply]
This may be a bug, most likely you are missing some dependencies, I've moved this to a new issue: #33

**Reply by MarkusJx** [reply]
Are you using a 32 or 64 bit system?

**Reply by tripower** [reply]
Windows is 64bit but I need to build it for 32bit
because of other parts 
can I do something else or only wait for solving of the issue 33?

@MarkusJx 
can you reproduce the problems if you use the same "products" and versions?

and thank you very much for the quick replies and help!

greetings
Klaus

**Reply by MarkusJx** [reply]
This will be fixed in `v2.1.7`, the issue originated from napi-rs not handling win32 very well: https://github.com/napi-rs/napi-rs/issues/297.
As this issue was fixed in there some time ago, this was *technically* fixed, but I've generated this project on top of a bugged version so this bug was still present.

I'll create a new release probably tomorrow and will inform you once that's done

**Reply by tripower** [reply]
thx

hope I can test it soon and 
than I will give feedback
fingerscrossed🙏

greetings
Klaus

**Reply by MarkusJx** [reply]
Version `2.1.7` is out now, I've tested it on win32 and it seems to work fine

**Reply by tripower** [reply]
cool thx
real fast help and build!!!
my test should by on monday
I will inform you about the state

greetings and nice weekend

**Reply by tripower** [reply]
hello @MarkusJx 
good news, it is working 
thanks for the quick bugfix

greetings Klaus
